examples/vision/metric_learning.py

```diff
@@ -0,0 +1,316 @@
+    return collage
+
+
+# Show a collage of 5x5 random images.
+sample_idxs = np.random.randint(0, 50000, size=(5, 5))
+examples = x_train[sample_idxs]
+show_collage(examples)
+
+"""
+Metric learning provides training data not as explicit `(X, y)` pairs but instead uses
+multiple instances that are related in the way we want to express similarity. In our
+example we will use instances of the same class to represent similarity; a single
+training instance will not be one image, but a pair of images of the same class. When
+referring to the images in this pair we'll use the common metric learning names of the
+`anchor` (a randomly chosen image) and the `positive` (another randomly chosen image of
+the same class).
+"""
+
+"""
+To facilitate this we need to build a form of lookup that maps from classes to the
+instances of that class. When generating data for training we will sample from this
+lookup.
+"""
+
+class_idx_to_train_idxs = defaultdict(list)
+for y_train_idx, y in enumerate(y_train):
+    class_idx_to_train_idxs[y].append(y_train_idx)
+
+class_idx_to_test_idxs = defaultdict(list)
+for y_test_idx, y in enumerate(y_test):
+    class_idx_to_test_idxs[y].append(y_test_idx)
+
+"""
+For this example we are using the simplest approach to training; a batch will consist of
+`(anchor, positive)` pairs spread across the classes. The goal of learning will be to
+move the anchor and positive pairs closer together and further away from other instances
+in the batch. In this case the batch size will be dictated by the number of classes; for
+CIFAR-10 this is 10.
+"""
+
+num_classes = 10
+
+
+class AnchorPositivePairs(keras.utils.Sequence):
+    def __init__(self, num_batchs):
+        self.num_batchs = num_batchs
+
+    def __len__(self):
+        return self.num_batchs
+
+    def __getitem__(self, _idx):
+        x = np.empty((2, num_classes, height_width, height_width, 3), dtype=np.float32)
+        for class_idx in range(num_classes):
+            examples_for_class = class_idx_to_train_idxs[class_idx]
+            anchor_idx = random.choice(examples_for_class)
+            positive_idx = random.choice(examples_for_class)
+            while positive_idx == anchor_idx:
+                positive_idx = random.choice(examples_for_class)
+            x[0, class_idx] = x_train[anchor_idx]
+            x[1, class_idx] = x_train[positive_idx]
+        return x
+
+
+"""
+We can visualise a batch in another collage. The top row shows randomly chosen anchors
+from the 10 classes, the bottom row shows the corresponding 10 positives.
+"""
+
+for examples in AnchorPositivePairs(num_batchs=1):
+    pass
+
+show_collage(examples)
+
+"""
+## Embedding model
+
+We define a custom model with a `train_step` that first embeds both anchors and positives
+and then uses their pairwise dot products as logits for a softmax.
+"""
+
+
+class EmbeddingModel(keras.Model):
+    def train_step(self, data):
+        # Note: Workaround for open issue, to be removed.
```

Comment:
Another thing to consider would be to have your generator return a dict `{"anchors": ..., "positives": ...}`, that way you don't need this and your code will be more explicit

Reply:
looks like would stil need the workaround though; in that case it returns a single element tuple, with the dictionary in it